Windows static library build fails with LNK2005: tls_callback_func already defined

Greetings! 

As we are updating the CycloneDDS version in [Conan Center](https://github.com/conan-io/conan-center-index/pull/30113) to package 11.0.1, during the development and review was noted a failure when trying to build the project as a static library on Windows, using MSVC compiler + `BUILD_SHARED_LIBS=OFF`: 

The build fails during the link step of `cycloneddsidl.dll` (`src/idl/idl.vcxproj`) with:

```
ddsc.lib(cdtors.obj) : error LNK2005: tls_callback_func already defined in string.obj
cycloneddsidl.dll : fatal error LNK1169: one or more multiply defined symbols found
Done Building Project "...\build\src\idl\idl.vcxproj" (default targets) -- FAILED.

Build FAILED.
```

You can take a look in full build log with the error: [cyclonedds-11.0.1-static.log](https://github.com/user-attachments/files/29737927/cyclonedds-11.0.1-static.log)

On the other hand, when building as shared library, the mentioned error does not occur and we can build with success. You can check the full build log with the shared library as well: [cyclonedds-11.0.1-shared.log](https://github.com/user-attachments/files/29738265/cyclonedds-11.0.1-shared.log)

There are a few indications this is a known problem, but not explicit issue reporting the case, I can list:
- #684 - Closed as fixed, but shows similar error
- #1313 - Closed as fixed, but presents the same error as for 11.0.1
- #2298 touches the mentioned problem, but lacks to express the exact problem and explain how that change fixes the case.
- The commit https://github.com/eclipse-cyclonedds/cyclonedds/blob/e2df2c305154ffc02cff0817cce99ba3c7b8978b/.github/workflows/windows.yml#L62 indicates there is a limitation regarding static on Windows

The error occurs because both `src/ddsrt/src/cdtors.c` and `src/idl/src/string.c` define the same global symbol `tls_callback_func`:

https://github.com/eclipse-cyclonedds/cyclonedds/blob/11.0.1/src/ddsrt/src/cdtors.c#L200

```c
#pragma comment (linker, "/INCLUDE:tls_callback_func")
EXTERN_C const PIMAGE_TLS_CALLBACK tls_callback_func = ddsrt_cdtor;
```

https://github.com/eclipse-cyclonedds/cyclonedds/blob/11.0.1/src/idl/src/string.c#L431

```c
#pragma comment (linker, "/INCLUDE:tls_callback_func")
EXTERN_C const PIMAGE_TLS_CALLBACK tls_callback_func = idl_cdtor;
```

This is fine for a shared build as each DLL carries its own copy. The problem arises because https://github.com/eclipse-cyclonedds/cyclonedds/blob/11.0.1/src/idl/CMakeLists.txt#L155 unconditionally declares the `idl` target as `SHARED`:

```cmake
add_library(idl SHARED ${headers} ${sources})
```

As a result, even when using `-DBUILD_SHARED_LIBS=OFF`, `cycloneddsidl.dll` is still built as a shared library, and links against the static `ddsc.lib` (which has `cdtors.obj` baked in), causing the linker to encounter `tls_callback_func` from both `cdtors.obj` and `string.obj` in the same link unit.

Note the same error still occurs on the master branch: [cyclone-master-static.log](https://github.com/user-attachments/files/29739200/cyclone-master-static.log)

#### Environment
* CycloneDDS: 11.0.1
* OS: Windows 10
* CPU: x86_64
* Compiler: MSVC 19.44.35228.0 (Visual Studio 2022, v17) |
* CMake: 4.3.2 with generator "Visual Studio 17 2022"


#### Steps to Reproduce

```
wget https://github.com/eclipse-cyclonedds/cyclonedds/archive/11.0.1.tar.gz
tar zxf 11.0.1.tar.gz && cd cyclonedds-11.0.1/
cmake -S . -B build -DEXPORT_ALL_SYMBOLS=OFF  -DINSTALL_PDB=OFF -DCYCLONEDDS_BUILD_TESTING=OFF -DBUILD_SHARED_LIBS=OFF -DENABLE_SECURITY=OFF -DBUILD_TESTING=OFF -DCMAKE_VERBOSE_MAKEFILE=ON
cmake --build build --config=Release
```

I could open a PR fixing only this point in case you are not pushing #2298 at this time. Just let me know!

--- a/src/ddsrt/src/cdtors.c
+++ b/src/ddsrt/src/cdtors.c
@@ -215,15 +215,15 @@ DDSRT_WARNING_CLANG_ON(missing-prototypes)
   PIMAGE_TLS_CALLBACK __crt_xl_tls_callback__ __attribute__ ((section(".CRT$XLY"))) = ddsrt_cdtor;
 #elif _WIN64
   #pragma comment (linker, "/INCLUDE:_tls_used")
-  #pragma comment (linker, "/INCLUDE:tls_callback_func")
+  #pragma comment (linker, "/INCLUDE:tls_callback_func_ddsrt")
   #pragma const_seg(".CRT$XLY")
-  EXTERN_C const PIMAGE_TLS_CALLBACK tls_callback_func = ddsrt_cdtor;
+  EXTERN_C const PIMAGE_TLS_CALLBACK tls_callback_func_ddsrt = ddsrt_cdtor;
   #pragma const_seg()
 #else
   #pragma comment (linker, "/INCLUDE:__tls_used")
-  #pragma comment (linker, "/INCLUDE:_tls_callback_func")
+  #pragma comment (linker, "/INCLUDE:_tls_callback_func_ddsrt")
   #pragma data_seg(".CRT$XLY")
-  EXTERN_C PIMAGE_TLS_CALLBACK tls_callback_func = ddsrt_cdtor;
+  EXTERN_C PIMAGE_TLS_CALLBACK tls_callback_func_ddsrt = ddsrt_cdtor;
   #pragma data_seg()
  #endif
 #else /* _WIN32 */

--- a/src/idl/src/string.c
+++ b/src/idl/src/string.c
@@ -426,15 +426,15 @@ _Pragma("GCC diagnostic pop")
   PIMAGE_TLS_CALLBACK __crt_xl_tls_callback__ __attribute__ ((section(".CRT$XLY"))) = idl_cdtor;
 #elif defined _WIN64
   #pragma comment (linker, "/INCLUDE:_tls_used")
-  #pragma comment (linker, "/INCLUDE:tls_callback_func")
+  #pragma comment (linker, "/INCLUDE:tls_callback_func_idl")
   #pragma const_seg(".CRT$XLY")
-  EXTERN_C const PIMAGE_TLS_CALLBACK tls_callback_func = idl_cdtor;
+  EXTERN_C const PIMAGE_TLS_CALLBACK tls_callback_func_idl = idl_cdtor;
   #pragma const_seg()
 #else
   #pragma comment (linker, "/INCLUDE:__tls_used")
-  #pragma comment (linker, "/INCLUDE:_tls_callback_func")
+  #pragma comment (linker, "/INCLUDE:_tls_callback_func_idl")
   #pragma data_seg(".CRT$XLY")
-  EXTERN_C PIMAGE_TLS_CALLBACK tls_callback_func = idl_cdtor;
+  EXTERN_C PIMAGE_TLS_CALLBACK tls_callback_func_idl = idl_cdtor;
   #pragma data_seg()
 #endif /* _WIN32 */
 

